Add ability to specify commit message on the command line

kbilsted opened the pull request:
This PR extends the command line commit command to enable a commit message. 

I was a bit in doubt whether we should include a new field in `FormCommit.cs` as I have done, or simply modify the `MergeOrCommitMessage` on the `_commitMessageManager`. The later will save the message to disc though and this is why I opted out of the idea

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

pmiossec commented:
Strange.

 What is the use case?
Couldn't you just use the 'git' command line?

kbilsted commented:
Well, you can say that about most operations in gitextensions ;)

Recall The commit command only opens the commit window where you then can select files and do hunk selections. It does not do a commit. I simply extended it to pre-fill the message

For my case a script will change files and build a commit message but i still like to do manual review in my favorite git tool before comitting

RussKie commented:
@kbilsted could you please open a new issue and describe the use case.
Like @pmiossec I am also a bit unclear what you are trying to achieve.
Thank you

gerhardol commented:
The pr is not more confusing than the ge cmd line commit itself. It could be used in a semi automated flow. Add an issue.

kbilsted commented on GitUI/CommandsDialogs/FormCommit.cs:
```diff
@@ -180,6 +180,7 @@ public sealed partial class FormCommit : GitModuleForm
         [CanBeNull] private IReadOnlyList<GitItemStatus> _currentSelection;
         private int _alreadyLoadedTemplatesCount = -1;
         private EventHandler _branchNameLabelOnClick;
+        private string _textFromCommandLine;
```
good idea

kbilsted commented on GitUI/Resources/ChangeLog.md:
```diff
@@ -1,6 +1,8 @@
 Changelog
 =========
 
+* The commit command (command line) accepts a commit message as an argument
```
so I should remove it?

mstv replied:
yes

kbilsted commented:
I have addressed the comments and rebased onto newest master. 

I've also changed the command line slightly to require a `--message` since that is how the command line parser is designed.

Finally, I've implemented that commit messages from the command line is **not** stored when the form is closed since I figured that if you close the form, you are not interested in committing, and ie. not interested in keeping the commit message. If you are, you can invoke from the commandline again.

mstv commented on GitCommands/CommitMessageManager.cs:
```diff
@@ -36,22 +37,24 @@ public sealed class CommitMessageManager : ICommitMessageManager
+        [CanBeNull]
+        private string _overriddenCommitMessage;
         private Encoding _commitEncoding;
         private IFileSystem _fileSystem;
```
Nitpick: In new files like this, we use alphabetical order for class members of the same category.

kbilsted replied:
No problem. But I don't understand your request. Where do you want me to put it? its defined after "m" which is alphabetically ordered already.

mstv replied:
`private readonly` is one category `private [non-readonly]` another.
Each category should be sorted alphabetically.
I.e. move the declaration below `_fileSystem`.

kbilsted replied:
thanks

mstv commented on GitUI/GitUICommands.cs:
```diff
@@ -493,7 +493,7 @@ bool Action()
                     }
 
-                    using (var form = new FormCommit(this))
+                    using (var form = new FormCommit(this, commitMessage: commitMessage))
```
Nitpick: I would not add name the argument if the given variable has a similar name.

mstv replied:
The final nitpick: Remove `commitMessage: `, please. Everything else is in a good shape.

kbilsted replied:
Not sure why my previous comment on the matter is gone. I can't remove it. I get the following error if I do `Severity	Code	Description	Project	Line	File	Suppression State
Error	CS1503	Argument 2: cannot convert from 'string' to 'GitUI.CommandsDialogs.CommitKind'	GitUI	496	C:\src\gitextensions\GitUI\GitUICommands.cs	Active
` due to multiple optional parameters with default values

mstv replied:
OK. With naming the argument `commitMessage`, the compiler can chose the default value for second argument `commitKind`.

RussKie commented:
@kbilsted please address @mstv's comments and we'll have it merged.
Thank you

gerhardol commented:
@kbilsted You have to contribute something more universally useful after merging this...

kbilsted commented:
@gerhardol what do you have in mind?

gerhardol commented:
> what do you have in mind?

Nothing specific, I like useful features and corrections
This is a feature with very limited usage. I can live with the "noise" this creates, implementation looks OK

If you took this as a "threat" to reject, then I had a little fun.

kbilsted commented:
I was so scared mom had to bring hot milk and biscuits

mstv commented:
@msftbot merge in 10 hours
